src/VisualStudio/Core/Def/Implementation/ProjectSystem/RunningDocumentTableEventTracker.cs

```diff
@@ -0,0 +1,192 @@
+            // OnAfterAttributeChangeEx for the temporary buffer. The native IDE used this even to
+            // add misc files, so we'll do the same.
+            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_DocDataReloaded) != 0)
+            {
+                if (CheckPreconditions(docCookie))
+                {
+                    _listener.OnReloadDocumentData(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie));
+                }
+            }
+
+            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_Hierarchy) != 0)
+            {
+                _listener.OnRefreshDocumentContext(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie));
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnBeforeDocumentWindowShow(uint docCookie, int fFirstShow, IVsWindowFrame pFrame)
+        {
+            if (fFirstShow != 0)
+            {
+                if (TryGetBuffer(docCookie, out var buffer))
+                {
+                    _listener.OnBeforeOpenDocument(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie), buffer);
+                }
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterDocumentWindowHide(uint docCookie, IVsWindowFrame pFrame)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnBeforeSave(uint docCookie)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        /// <summary>
+        /// Gets the text buffer for a document cookie.
+        /// Also checks to make sure the document is initialized before returning.
+        /// </summary>
+        public bool TryGetBuffer(uint docCookie, out ITextBuffer textBuffer)
+        {
+            textBuffer = null;
+            if (!CheckPreconditions(docCookie))
+            {
+                return false;
+            }
+
+            if ((object)_runningDocumentTable.GetDocumentData(docCookie) is IVsTextBuffer bufferAdapter)
+            {
+                textBuffer = _editorAdaptersFactoryService.GetDocumentBuffer(bufferAdapter);
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Checks that we're on the UI thread and that the document has already been initialized.
+        /// </summary>
+        private bool CheckPreconditions(uint docCookie)
+        {
+            _foregroundAffinitization.AssertIsForeground();
+
+            if (!_runningDocumentTable.IsDocumentInitialized(docCookie))
+            {
+                // We never want to touch documents that haven't been initialized yet, so immediately bail. Any further
+                // calls to the RDT might accidentally initialize it.
+                return false;
+            }
+
+            return true;
+        }
+
+        #region IDisposable Support
+        private void Dispose(bool disposing)
```

Comment:
Isn't this pattern (having the "disposing") usually for if we're doing a finalizer as well?

Reply:
Good question - this is also an area I'm not super familiar with!  I guess it looks like I should be using the Dispose() pattern?  https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose